cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,1378 @@ func TestCreateVPCSubnet(t *testing.T) {
 		g.Expect(err).ToNot(BeNil())
 	})
 }
+
+func TestReconcileVPCSecurityGroups(t *testing.T) {
+	var (
+		mockVPC  *mock.MockVpc
+		mockCtrl *gomock.Controller
+	)
+
+	securityGroupID := "securityGroupID"
+	securityGroupName := "securityGroupName"
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When Security group ID is set and returns error while getting security group", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			ID: ptr.To("securityGroupID"),
+		}
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroup(gomock.Any()).Return(nil, nil, errors.New("failed to get security group"))
+		err := clusterScope.ReconcileVPCSecurityGroups()
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When Security group ID is set and security group doesn't exist", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			ID: ptr.To("securityGroupID"),
+		}
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroup(gomock.Any()).Return(nil, nil, nil)
```

Comment:
I think `GetSecurityGroup` won't return nil, if we pass a id to look for, it would return an error, please verify this behaviour once with cloud API call and modify the TC accordingly.